Add UIForm Element

GimLala opened the pull request:
Auto resizing containers added (Documentation on website is remaining)

These containers automatically resize when elements are added to them.

I was having some problems running tests, but I had tested them separately in a project I was using them in. More similar elements are coming like `UIAutoScrollingContainer `which uses an auto resizing container as its scrollable container instead of a normal `UIContainer`.
Check if you like this implementation, or maybe we could even merge `UIContainer `and the new `UIAutoResizingContainer`

**Adding PR split list to first post:**

**Easy changes**
1.  [x] Adding .idea to .gitignore PR - #530

**Changes to existing base classes**
1. [x] Changes being made to UIElement - This needs some merge effort with other recent changes, I had a stab locally but I think it can probably be done better. Also I need to interrogate what theses change are more fully as changes here can have a wide impact across the codebase as I discovered last week with a smaller change to dynamic resizing, that passed all the unit tests, managing to break an image resizing demo. #531, #534, #535, #536
2. [x] Changes to UIContainer - Same reasoning as above regarding wide ranging impact. #537
3. [x] Changes to UIScrollingContainer - I'm less concerned about changes here as it is not widely used, but it's worth looking at - there is also the possibility we may want to merge this class entirely with AutoScrollingContainer. #538, #539

**New stuff**
1. [x] AutoScrollingContainer #545
2. [x] AutoResizingContainer #543, #544
3. [x] UI2DSlider #540
4. [ ] UIForm

**Using the new stuff in existing stuff**
1. [x] UIColourPicker addition of UI2DSlider #542

MyreMylar commented:
The main question I had when looking over this is - are there circumstances where you would want a container that resizes like this, as opposed to one that resizes it's scrollable area (e.g. the way an explorer or one where you have decided the size ahead of time to fit the stuff you are going to put inside of it? Most UIs I see these days adopt a 'squish and vertical scroll' for oversized content - I suspect mainly because of the dominance of phones and tablets in browsing content. 

I suspect there are - but listing off some ideas as to what this would be used for in a UI would probably help clarify the design and help create an example that uses this type of container for the examples repository:

https://github.com/MyreMylar/pygame_gui_examples

GimLala commented:
I'll admit that I didn't have too much in mind except it would be easy to replace the container used in the UIScrollableContainer to this special container. I thought we would have to implement something like this either straight into the UIScrollableContainer otherwise anyways, and implementing it this way would potentially allow for more use cases to be available with the same effort for maintainence.

MyreMylar commented on pygame_gui/elements/ui_auto_resizing_container.py:
```diff
@@ -0,0 +1,258 @@
+        """
+        super().update(time_delta)
+        # Perhaps this could be attached to a timer to not have to call every single frame
+        self._update_anchor_elements()
+        self._update_sorting()
+        self._update_extreme_elements()
+        self._update_dimensions()
```
Instead of doing this - checking the container's contents every loop (which could have large performance implications in a container with lots of elements inside), I think we should update these things only when we know the container may need to change size.

As far as I can tell this is in one of four situations:

1. A new element has been added to the container.
2. An existing element has been removed form the container.
3. An element already inside the container has changed position.
4. An element within the container has changed size.

There is already a function on all containers that all elements call when they change size or move, it is called `on_anchor_target_changed()`. I would rename this function in the `UIContainer()` to something more generic like `on_content_moved_or_resized()` and then override it in this class and have it do this work.

For adding and removing you can just override the existing `add_element())` and `remove_element()`. Once you've covered all four you should be able to remove this update function I would think?

MyreMylar commented on pygame_gui/elements/ui_auto_resizing_container.py:
```diff
@@ -0,0 +1,258 @@
+            left = min(left, self.left_element.get_abs_rect().left)
+        if self.max_edges_rect is not None:
+            left = max(left, self.max_edges_rect.left)
+        return left
+
+    def get_right_most_point(self) -> int:
+        """
+        Gets the maximum x value any element has clamped to the min_edges_rect and max_edges_rect
+        :return: An integer
+        """
+        right = self.min_edges_rect.right
+        if self.right_element:
+            right = max(right, self.right_element.get_abs_rect().right)
+        if self.max_edges_rect is not None:
+            right = min(right, self.max_edges_rect.right)
+        return right
+
+    def get_top_most_point(self) -> int:
+        """
+        Gets the minimum y value any element has clamped to the min_edges_rect and max_edges_rect
+        :return: An integer
+        """
+        top = self.min_edges_rect.top
+        if self.top_element:
+            top = min(top, self.top_element.get_abs_rect().top)
+        if self.max_edges_rect is not None:
+            top = max(top, self.max_edges_rect.top)
+        return top
+
+    def get_bottom_most_point(self) -> int:
+        """
+        Gets the maximum y value any element has clamped to the min_edges_rect and max_edges_rect
+        :return: An integer
+        """
+        bottom = self.min_edges_rect.bottom
+        if self.bottom_element:
+            bottom = max(bottom, self.bottom_element.get_abs_rect().bottom)
+        if self.max_edges_rect is not None:
+            bottom = min(bottom, self.max_edges_rect.bottom)
+        return bottom
```
Are these for internal use? I feel like for the end user they just want to be doing:

`my_container.rect.bottom` or `my_container.rect.top` and have all this already precalculated for them. Perhaps I have misunderstood the purpose.

If they are for internal use then they should also have the `_` underscore prefix.

MyreMylar commented on pygame_gui/elements/ui_auto_resizing_container.py:
```diff
@@ -0,0 +1,258 @@
+
+    def _update_extreme_elements(self) -> None:
+        """
+        Updates which elements is currently responsible for preventing the container to collapse from a certain side
```
```suggestion
        Updates which elements are currently responsible for preventing the container from collapsing on each side
```

MyreMylar reviewed (dismissed):
Ok, I've finally had a chance to look over this. Apologies for the delay.

I noticed while testing this that something in the rendering machinery is not ready for:

```
test_resizing_container = UIAutoResizingContainer(pygame.Rect(0, 0, 50, 50))
hello_button = UIButton((60, 0), 'Hello', container=test_resizing_container)
```
It seems to get negative dimensions passed down to it and is unhappy, so that is something to look into further.

The `on_content_moved_or_resized()` function I suggested for the UIContainer may also be useful when trying to use this in the `UIScrollingContainer` as we will need to trigger the code that is currently in the `UIScrollingContainer`'s `set_scrollable_area_dimensions()` function somehow, perhaps we can have it set a flag variable 'this container has resized` that the `UIScrollingContainer` can check.

Anyway, thanks for your hard work on this. I think it is getting somewhere useful 🎉

GimLala commented:
Hi,
Thank you for reviewing the pull request. I also found some flaws with this code when resizing the container. I think there are certain other major problems with the class which I am trying to pin point properly.
However, I had already finished most work on the uiscrolling container which automatically resizes, and some other elements.
Yet I cannot guarantee that I will have the time to fix all the issues as I have some important exams coming up which will last the whole of February. I will try my best to debug the problems and update the necessary code, but I think it might be better to have the other elements like:
* UIAutoScrollingContainer
* UI2DSlider
* UIColourPickerDialog (with an added 2Dslider in the colour square which allows dragging to select the correct colour) 
* UIForm

These elements aren't fully completed, and I'm sure there are bugs but you'll be able to see what my approach to those elements is and guide me if necessary. You may even finish it if I am too inactive. It will atleast get the process of adding those elements going

GimLala commented:
While renaming the `UIContainer` method `on_anchor_target_changed` to `on_content_moved_or_resized` as you suggested, I saw that the UIElement class has a method called `_on_content_changed` which has no documentation. What exactly does this method do? Perhaps some documentation with that would make it unambiguous about which function does what.

Secondly, the `get_top_most_point` is not necessarily the same as getting the `self.rect.top`, because `self.rect.top` will not be changed to the y value from `get_top_most_point` unless `resize_top` is `True`. This means that using this method could be way to check what size the container will be in the case where `resize_top` is `False` and they wish to toggle it to `True`.
However, I can add the prefix if you feel like that use case is too specific and may confuse the users.

GimLala commented:
I made some changes using your suggestions. However, I can just not get the container to work properly. It seems to be having some problems with buttons and their texts, and especially with dynamic width and heights.

GimLala commented:
I think I found the root of the problem. So what happens when we create a new button with dynamic width and height is:
1. First, it creates a full sized image containing the text.
2. Then it sees how much of the image fits in the container, and crops the image accordingly. 
3. *However, it also resizes its rect to the size of the image.*
4.  This means that since the AutoResizingContainer checks whether it needs to expand using the rect of the button, it fails to recognise that the button exceeds the borders.
5.  Now since the container doesn't resize, no matter how many times you try to resize the button, the button will always be invisible since the rect doesn't change, the clipping rect doesn't change and the image size doesn't change.

So, to fix this problem, we'll have to **stop the image from resizing based on if it's outside the container**. Just update the clipping rect instead and as soon as the element moves update the clipping rect, and only blit the the image actually contained within the container.

GimLala commented:
I finally was able to fix the problem by eleminating the if condition in `UIElement._set_image` which checks whether the `image_clipping_rect` is 0 and sets the `universal_emtpy_surface` as the `self.image`.

MyreMylar commented:
Looks like a crop of the unit tests are failing now I've enabled them to run.

GimLala commented:
> Looks like a crop of the unit tests are failing now I've enabled them to run.

I have tried fixing some of the failing tests with the next commit, could you try re-running the workflow? it doesn't allow me to run it myself

MyreMylar commented:
> > Looks like a crop of the unit tests are failing now I've enabled them to run.
> 
> I have tried fixing some of the failing tests with the next commit, could you try re-running the workflow? it doesn't allow me to run it myself. Also, when I run the tests in my IDE, many of them fail because the manager can't find the specified files, or the colours of the gradient don't match, etc. I don't think I have changed any code which would interfere with the colours or location of files. There is a test which specifically tests for the if condition I had removed, where in the expected image was supposed to be the `ui_manager.universal_empty_surface` but after I changed it (so that the size of the image isn't 0, 0) the tests fails.
> 
> I will also add the appropriate tests in a few days

I have kicked off the tests.

To run them locally you should just be able to run `pytest tests` from the root directory (the one that has the 'tests', 'docs' and 'pygame_gui' folders. For me it looks like this:

![image](https://github.com/MyreMylar/pygame_gui/assets/13382426/82a4ce19-915d-4be0-b73c-3ca23bfc108b)

GimLala commented:
Okay, I have finally seemed to squash the bugs coming up with the containers while resizing. Also, I added some other functions and removed other bugs with this, namely:
1. `UIContainer` now also has `expand_top` and `expand_left` functions which work like `set_dimensions` except, as the name suggests, they expand the container from the top and left respectively. The big difference between just setting the dimensions and moving the rect, and these functions is that these functions also update the relative rects of the contained elements. This means that none of the contained elements' absolute positions will move.
2. I updated the `on_contained_elements_changed` method to recursively update targets as when there was a chain of elements all anchored to each other, it would not update all of them previously

I ran all the tests and it seems to be working properly. I tested the elements manually and they are working but I have not written the pytest equivalent tests yet. It's a bit daunting of a task because there are so many things to test

GimLala commented:
This is pretty much all the additions I wanted to make in terms of elements and features. It's just the matter of writing the tests now, but I think you should check out all the changes first to make sure this is what you want

MyreMylar commented:
Looks like things are coming along. These are the failing tests at the minute:

```
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_creation - AttributeError: 'NoneType' object has no attribute 'width'
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_create_too_small - AttributeError: 'NoneType' object has no attribute 'width'
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_press_cancel_button - AttributeError: 'NoneType' object has no attribute 'width'
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_press_ok_button - AttributeError: 'NoneType' object has no attribute 'width'
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_click_in_saturation_value_square_button - AttributeError: 'NoneType' object has no attribute 'width'
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_mess_with_colour_channel_event - AttributeError: 'NoneType' object has no attribute 'width'
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_update_saturation_value_square - AttributeError: 'NoneType' object has no attribute 'width'
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_update_current_colour_image - AttributeError: 'NoneType' object has no attribute 'width'
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_changed_rgb_update_hsv - AttributeError: 'NoneType' object has no attribute 'width'
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_changed_hsv_update_rgb - AttributeError: 'NoneType' object has no attribute 'width'
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_show - AttributeError: 'UI2DSlider' object has no attribute 'sliding_button'
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_hide - AttributeError: 'NoneType' object has no attribute 'width'
FAILED tests/test_windows/test_ui_colour_picker_dialog.py::TestUIColourPickerDialog::test_show_hide_rendering - AttributeError: 'UI2DSlider' object has no attribute 'sliding_button'
```
And most of them look like the same fail there.

GimLala commented:
Ah, My "Fixed Potential Bug" commit is the culprit here. Added comments so that I don't unfix a bug again. However, 2 of the tests are still failling I'll see what is causing them

GimLala commented:
The tests were failing mainly because the `UIButton` in the `UI2DSlider` kept consuming the event so the colour wouldn't change. Usually this wouldn't be a problem because clicking the button in place shouldn't change the colour anyway, but because this was done by the program where the colour was set programmatically and the necessary values not updated properly, it would cause errors

GimLala commented:
One of the tests is failing but I can't figure out why.

GimLala commented:
Why are the two scrollbars implemented a bit differently? could be something to do with it. I think I need your help debugging with this one, @MyreMylar because I can't even figure out what the bug is and whether it affects the container at all to the user

GimLala commented:
Aha! I have fixed the bugs. It had to do with the visible container being resized after calculations of the sizes of the scrollbars instead of before the calculations, as it was done before. But I think I've really made all the changes I want and just need to create the tests.

MyreMylar commented:
Great looks like the current tests are compiling! 

I will give this a good look over today, if its all good then we'll just need some new unit tests to cover the new elements.

MyreMylar commented:
Absolutely, did not get onto looking at this today sorry, but it is on my to do list for this week.

MyreMylar commented:
OK, I've got this running locally and browsed through bits of it - I have run into a few bugs along the way, most of which should get squished through coverage compliant unit testing - which I'll add some separate comments about as I go.

I think my biggest issue overall with this PR is its size is making it tough to mentally digest everything going on inside of it. Sadly I only have a small amount of time each week to work on coding projects at the moment which makes dumping this much code in my head at once very tough!

I'm going to suggest we break this down into about nine PRs and in the following order - like so:

**Easy changes**
1. Adding .idea to .gitignore PR - #530

**Changes to existing base classes**
2. Changes being made to UIElement - This needs some merge effort with other recent changes, I had a stab locally but I think it can probably be done better. Also I need ot interrogate what theses change are more fully as changes here can have a wide impact across the codebase as I discovered last week with a smaller change to dynamic resizing, that passed all the unit tests, managing to break an image resizing demo.
3. Changes to UIContainer - Same reasoning as above regarding wide ranging impact.
4. Changes to UIScrollingContainer - I'm less concerned about changes here as it is not widely used, but it's worth looking at - there is also the possibility we may want to merge this class entirely with AutoScrollingContainer.

**New stuff**
5. AutoScrollingContainer
6. AutoResizingContainer
7. UI2DSlider
8. UIForm

**Using the new stuff in existing stuff**
9. UIColourPicker addition of UI2DSlider

Then we can make sure each step is properly looked at and tested before moving up the chain. I'll help with this task by creating PRs with the changes and adding you as co-author to the commits so you get credit. You can also join in this process if you like, not sure how much you've used git and multiple branches, but I'll start off with the first couple anyway.

MyreMylar commented on pygame_gui/elements/ui_form.py:
```diff
@@ -0,0 +1,1144 @@
+
+                elif type_name == "boolean":
+                    element = UIDropDownMenu(options_list=["True", "False"],
+                                             starting_option=args.get("default", "True"),
```
```suggestion
                                             starting_option=str(args.get("default", "True")),
```

I had to add a cast here to get this code to run.

GimLala replied:
When you create a boolean with a default argument, you're supposed to do it like `boolean(default="True")` and not without quotes as in `boolean(default=True)`, due to UIDropDown not accepting values except strings. Infact I thought that was already enforced. But now I change my mind, and think that it makes more sense if we do the casting behind the scenes ourselves, and then when we need to get the values, re-cast it to boolean as UIDropDown will only return strings as it's current option.

MyreMylar commented on pygame_gui/elements/ui_form.py:
```diff
@@ -0,0 +1,1144 @@
+            if event.ui_element == self.submit_button:
+                form_values = self.get_current_values()
+
+                # old event - to be removed in 0.8.0
+                event_data = {"user_type": OldType(UI_FORM_SUBMITTED),
+                              "form_values": form_values,
+                              "ui_element": self,
+                              "ui_object_id": self.most_specific_combined_id}
+                pygame.event.post(pygame.event.Event(pygame.USEREVENT, event_data))
```
```suggestion
```
We don't need to add the old event type to new code, it's just there in older code for backwards compatibility.

MyreMylar commented on pygame_gui/elements/ui_form.py:
```diff
@@ -0,0 +1,1144 @@
+
+        # You can click this button to expand/hide the container
+        self.main_expand_button = UIButton(pygame.Rect(0, 0, 0, 0),
+                                           text=self.name + " ▼",
```
```suggestion
                                           text=self.name + " ▼",
```

Slightly annoyingly the new default font, while otherwise better than the old default font, doesn't support the same range of symbols. I kind of want to add HTML parsing to button text just to let us fix this use case by switching to the old default font midway through this text.

Anyway, right now it won't work unless we style this whole button using 'fira_code_symbols'.

MyreMylar commented on pygame_gui/elements/ui_form.py:
```diff
@@ -0,0 +1,1144 @@
+        consumed = False
+        if event.type == UI_BUTTON_PRESSED and event.mouse_button == pygame.BUTTON_LEFT:
+            if event.ui_element == self.submit_button:
+                form_values = self.get_current_values()
```
The output I get from this feels wrong:
```
{'section1': {'Question 1': <UILabel GUISprite(in 1 groups)>, 'Question 2': <UILabel GUISprite(in 1 groups)>}}
```

I feel like I should be getting the string/boolean values of those questions answers.

GimLala replied:
Yes! I had not made the necessary changes after adding the NamedTuple "InputField". It should be fixed now.

MyreMylar commented on pygame_gui/core/ui_container.py:
```diff
@@ -250,6 +250,7 @@ def expand_top(self, height_increase: int) -> None:
         This is achieved by setting the new dimensions and updating the anchors of all the elements anchored
         to the top of the container.
 
+        :param height: The height to increase by. Pass in negative values to decrease the size
```
```suggestion
```

MyreMylar commented:
OK, I've broken down most of the changes to the core stuff now into new PRs and added tests (the remaining changes are ones I'm not sure are actually needed yet).

I'm going to split off the 4 new elements into separate PRs next.

GimLala commented:
Hi, I have been out recently for vacation on a tour. I'll see what exactly are the changes when I get back in a few days. I realised the PR got very big very quickly so it's great that you're breaking it down. Some bugs I can tell were caused by short sighted changes by me failing to account for their impact, but they should be easily fixable.

MyreMylar commented on tests/test_elements/test_ui_auto_resizing_container.py:
```diff
@@ -24,6 +24,7 @@ def test_add_element(self, _init_pygame, default_ui_manager: IUIManagerInterface
                           manager=default_ui_manager)
         default_ui_manager.get_root_container().remove_element(button)
         container.add_element(button)
+
```
```suggestion
```

MyreMylar commented on tests/test_elements/test_ui_auto_resizing_container.py:
```diff
@@ -83,6 +84,7 @@ def test_remove_element(self, _init_pygame, default_ui_manager: IUIManagerInterf
         container.remove_element(button_bottom)
 
         container.update(0.4)
+
```
```suggestion
```

MyreMylar commented on tests/test_core/test_ui_element.py:
```diff
@@ -620,7 +620,8 @@ def test_set_image(self, _init_pygame, _display_surface_return_none, default_ui_
         element._set_image_clip(pygame.Rect(0, 0, 0, 0))
         element._set_image(coloured_surface_1)
-        assert element.image == default_ui_manager.get_universal_empty_surface()
+        assert element.image.get_size() == (50, 50)
+        assert element.image.get_at((10, 10)) == pygame.Color(0, 0, 0, 0)
```
```suggestion
        assert element.image == default_ui_manager.get_universal_empty_surface()
```

GimLala commented:
Hi, in the other separate PR's, have you made any changes to the code from this PR? If that's the case, then the only change needed will be to this PR to fix the errors you mentioned above (which I think I have fixed)

MyreMylar commented:
> Hi, in the other separate PR's, have you made any changes to the code from this PR? If that's the case, then the only change needed will be to this PR to fix the errors you mentioned above (which I think I have fixed)

If you look at the 'files changed' link at the top there you can see what stuff is going to get merged from this PR now.

Essentially the other PRs I made have stripped out and merged to main everything except the UIForm stuff which is now all that is left in this one.

Last time I looked at the `UIForm` I realised that the drop down really needs to support a dynamic width to work properly with something like this - where arbitrary data can be added to the drop down. I wanted to figure out a way to do that (it's tricky because drop downs start closed but you would want a dynamic width one to be wide enough to fit every item in the drop down list. 

I think the form also needs a good looking example of a form creating in the pygame_gui_examples Github project as it is a complicated element and is likely to be vulnerable to changes in lots of other areas so I'd like to have something to run an check before making releases. 

So basically I want to get those steps worked out before merging it. They may in turn throw up other things. I have a little list of things to look at over my next work holiday and getting this merged in is part of it - along with a new release of the library.

MyreMylar commented:
I guess the form could also use a test file to give it a good work out and make the test coverage check pass.

GimLala commented:
I saw that in one of the branch merges with main, you un-committed the changes from my "adding default 0 sized scrollbars" commit, which added those scrollbars to the UIScrollingContainer so that the elements in the UIForm could anchor to it. Is there a reason you did so? Because now the UIForm isn't working for me.

I think that it's better to add a scrollbar to the dropdown then make it dynamically sized, as it may not look great if it expanded to the length of the entire screen. Perhaps it would extend up to the length specified in the height limit, then add a scrollbar which allows you to scroll through the options.

I had a working example (although perhaps not good looking: I had no specific theming applied), which pretty much used all the features I have added. I can modify and add that in once we fix the issues with the default scrollbars.

I'll also create a test file (or add any missing test functions if you create one before me), but I will likely only have time from around 6th because I have a really busy week coming up.

MyreMylar commented:
> I saw that in one of the branch merges with main, you un-committed the changes from my "adding default 0 sized scrollbars" commit, which added those scrollbars to the UIScrollingContainer so that the elements in the UIForm could anchor to it. Is there a reason you did so? Because now the UIForm isn't working for me.

The scrolling container is made op of two containers - an 'outer' one that contains the 'inner' container and the scroll bar and an 'inner' container that contains the contents. The contents should be unaffected by whether there is a scrollbar in the outer container or not as they should only know about the inner container's dimensions.

If there is an issue with the resizing of the inner container and the effect on it's contents then we need to solve that bug not mess the scroll bar which should have no impact on the contents. 


> I think that it's better to add a scrollbar to the dropdown then make it dynamically sized, as it may not look great if it expanded to the length of the entire screen. Perhaps it would extend up to the length specified in the height limit, then add a scrollbar which allows you to scroll through the options.

The drop down already has a scroll bar for height, I was talking about dynamic sizing for width here (and I guess also font size height) not the number of items in a drop down.

> I had a working example (although perhaps not good looking: I had no specific theming applied), which pretty much used all the features I have added. I can modify and add that in once we fix the issues with the default scrollbars.
> 
> I'll also create a test file (or add any missing test functions if you create one before me), but I will likely only have time from around 6th because I have a really busy week coming up.

Great! I will get back to this and give it a more thorough going over but any example usage code you have you can just slap into this issue in a code block so I can have look.

GimLala commented:
> If there is an issue with the resizing of the inner container and the effect on it's contents then we need to solve that bug not mess the scroll bar which should have no impact on the contents.
Well although the current problem is because the contained elements are anchored to the scrollbar instead of the right edge of the container, adding the zero sized scrollbars reduced a lot of code in having to manually create, position and delete the scrollbar instead of using anchors to easily control that automatically.

For example, I was able to anchor the viewing container to the scrollbars so that you wouldn't have to resize it manually when scroll bars are added-it would resize since it is anchored to the scrollbars.

So I think adding the 0 sized scrollbars, should be a good move because it reduces the need for manual tampering with the container and scrollbars, and allows a potentially easy implementation of the what you described in #497  (by anchoring the scrolling container to either the viewing container or the scrollbar when it is locked to scrolling only in one dimension)

GimLala commented:
> The drop down already has a scroll bar for height, I was talking about dynamic sizing for width here (and I guess also font size height) not the number of items in a drop down.

Oh! That would be a great addition, perhaps also a way to add text wrapping so that when the text exceeds a certain length it just wraps to take another line. Although I think that might be one of the features which would not be too commonly used as drop down options are usually meant to be kept short in size.

GimLala commented:
> Great! I will get back to this and give it a more thorough going over but any example usage code you have you can just slap into this issue in a code block so I can have look.

```
import pygame
import pygame_gui
from pygame_gui import UI_FORM_SUBMITTED
from pygame_gui.elements import UIButton, UIAutoResizingContainer, UIAutoScrollingContainer, UIForm, UIDropDownMenu

pygame.init()

pygame.display.set_caption('Quick Start')
window_surface = pygame.display.set_mode((800, 600))

background = pygame.Surface((800, 600))
background.fill(pygame.Color('#000000'))

manager = pygame_gui.UIManager((800, 600))

test_rect = pygame.Rect(40, 50, 700, 500)

# buttons = []
# resizing_test = UIAutoScrollingContainer(pygame.Rect(40, 50, 700, 500), manager, True, False)
# hello_button = UIButton(relative_rect=pygame.Rect((650, 275), (100, 50)),
#                         text='Say Hello 1',
#                         manager=manager, container=resizing_test)
# buttons.append(hello_button)
#
# hello_button = UIButton(relative_rect=pygame.Rect((25, 75), (150, 25)),
#                         text='Say Hello' + str(int(hello_button.text[-1]) + 1),
#                         manager=manager, container=resizing_test,
#                         anchors={"left_target": hello_button, "top_target": hello_button})
# buttons.append(hello_button)

questionnaire = {
    "Just enter one character:": "character",
    "Perhaps some more:": "short_text",
    "Or go wild!": "long_text",
    "how about a section?": {"do you like it?": "boolean", "well, here, section-ception": {"do you like it?": "boolean", "well, here, section-ception": {"finally...": "integer"}}},
    "Enter a float:": "decimal",
    "You must enter something here:": "short_text('here', True)",
    "Persome more:": "long_text",
    "let's try a button": UIButton(pygame.Rect(100, 100, 100, 100), "button"),
    "Peaps some more:": "short_text",
    "OO dropdown!": UIDropDownMenu(["foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo"], "foo",
                                   pygame.Rect(100, 100, 100, 100), ),
    "rhaps some more:": "short_text"}
form_test = UIForm(test_rect, questionnaire, manager)

print(form_test.combined_element_ids)
print(form_test.get_container().combined_element_ids)
print(form_test.parsed_questionnaire["how about a section?"].combined_element_ids)
print(form_test.parsed_questionnaire["how about a section?"].parsed_questionnaire["well, here, section-ception"].combined_element_ids)


clock = pygame.time.Clock()
is_running = True
pressed = False
element = None

while is_running:
    time_delta = clock.tick(30) / 1000.0
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            is_running = False

        elif event.type == pygame_gui.UI_BUTTON_PRESSED:
            print(event.ui_element.most_specific_combined_id)
            pressed = False
            element = None

        elif event.type == UI_FORM_SUBMITTED:
            print(event.form_values)

        manager.process_events(event)

    manager.update(time_delta)

    window_surface.blit(background, (0, 0))
    pygame.draw.rect(window_surface, "Yellow", form_test.scrollable_container.rect, 1)
    pygame.draw.rect(window_surface, "Green", form_test._view_container.rect, 1)
    pygame.draw.rect(window_surface, "White", form_test.rect, 1)
    s_cont = form_test.parsed_questionnaire["how about a section?"]
    pygame.draw.rect(window_surface, "Blue", s_cont.rect, 1)
    pygame.draw.rect(window_surface, "Blue", s_cont.section_container.rect, 1)
    # Uncomment this to see the structure of the "parsed_questionnaire"
    # This shows a questionnaire when the values have been replaced by elements which are displayed on the screen
    # print(form_test.parsed_questionnaire)

    pygame.draw.rect(window_surface, "Orange", form_test.submit_button.rect, 1)
    for label in form_test.parsed_questionnaire.values():
        if isinstance(label, pygame_gui.elements.ui_form.InputField):
            pygame.draw.rect(window_surface, "Orange", label[1].rect, 1)
            pygame.draw.rect(window_surface, "Orange", label[2].rect, 1)

    manager.draw_ui(window_surface)
    pygame.display.update()
```

GimLala commented:
Also I quickly fixed UIForm to remove scrollbars from anchors and it works as expected. Don't know why I added them in the first place. I remember there used to be some bugs with the submit button but they seem to have disappeared.

GimLala commented:
Also, I think we should merge UIAutoScrollingContainer with UIScrollingContainer in a fashion which is similar to how it is done with some other text elements (passing -1 for the corresponding dynamic size in that axis). I could work on that in the next few days once I'm done with the changes highlighted previously.

GimLala commented:
Okay I fixed all the bugs I mentioned above with positioning and sizing the elements after theming elements are updated. Another "bug" I found was that UIDropDownMenu doesn't automatically update the positioning of it's `current_state.open_button` when it is resized or repositioned directly without calling update_position or equivalent functions on it. This seems somewhat intended as written in the doc string of those methods but it seems to me that it should be done by the drop down menu automatically as the user should not need to call a separate function.

GimLala commented:
Some tests do seem to be failing, I'll take a look

GimLala commented:
I also made one change where the scrollbars are generated only when the corresponding allow_scroll variables are set to True. But a problem is that this makes the allow_scroll variables currently read only (if we want to be able to change those variables, then we'll have to check whether they are changed and generate the scrollbars and un-anchor the scrollable_container etc.).

I made this change because it made sense to me that when, for example, allow_scroll_x is False, then we don't have a horizontal scroll bar at all ever. This may not be a strong enough argument, so the changes can be easily reverted if you wish.

P.S: Ignore the gui_test and gui_test2 files that I accidentally commit (and removed). I used them for testing elements on my own.

GimLala commented:
what happened to the code coverage tests?

MyreMylar commented:
> what happened to the code coverage tests?

I think they don't run if the regular tests fail.

MyreMylar commented on pygame_gui/elements/ui_auto_scrolling_container.py:
```diff
@@ -64,11 +64,13 @@ def __init__(self,
             resize_left = False
             resize_right = False
             anchors["right"] = "right"
+            anchors["right_target"] = self.vert_scroll_bar
```
```suggestion
```

MyreMylar commented on pygame_gui/elements/ui_auto_scrolling_container.py:
```diff
@@ -64,11 +64,13 @@ def __init__(self,
             resize_top = False
             resize_bottom = False
             anchors["bottom"] = "bottom"
+            anchors["bottom_target"] = self.horiz_scroll_bar
```
```suggestion
```

GimLala commented:
There seems to be problems with the vertical scrollbar not detecting that the UIButton inside a UIScrollingContainer is hovered due to not calling set_container_this_will_scroll (which is quite a long name, perhaps should be renamed to set_container_to_scroll?).
What I don't understand is the need for both `are_contents_hovered` and focus sets. Shouldn't the scroll bar just check if anything in its focus set is hovered and then decide whether it needs to handle the event?

MyreMylar commented on pygame_gui/elements/ui_auto_scrolling_container.py:
```diff
@@ -84,7 +86,7 @@ def __init__(self,
                                                                 object_id="#scrollable_container",
                                                                 class_id=None),
-                                                            anchors=anchors)
+                                                            anchors=anchors.copy())
```
```suggestion
                                                            anchors=anchors)
```

MyreMylar replied:
rollback unneeded changes to auto-scrolling container

GimLala commented:
Why did you remove the anchors from the scrollable container? I had added them so that the container gets resized appropriately when scrollbars are "removed"

GimLala commented:
I think there might be a better way to handle vertical scrollbars' scrolling using focus sets but I think that should be in a different PR so I'll work on that after UIForm is done. Currently I'm working on tests for UIForm (especially testing the questionnaire-validation methods)

MyreMylar commented:
> Why did you remove the anchors from the scrollable container? I had added them so that the container gets resized appropriately when scrollbars are "removed"

I haven't? As far as I can tell, the scrolling container version in the main branch is basically the same as the one you had in here earlier today. 

The only changes I had to make were merging a recent fix from main that got lost in your refactor and clearing out an anchor target of the scrolling container to the scrollbars that didn't make much logical sense as the scrollable container is already only visible inside of the view container anyway - which is itself already sized by the scrollbars. We can just use the dimensions of the view container to anchor the scrollable container instead in this scenario.

I think there is still some work to do somewhere here as something in the setup for the UIForm is making the mousewheel scroll events not work again but at least the current tests are passing.

GimLala commented:
Oh I meant the anchor targets not the anchors. And yes I just realized they were unnecessary as I forgot the scrollable container is contained within the _view_container not the root container

MyreMylar commented:
OK, I can't find any obvious functionality bugs with the example code now.

That example still doesn't look great at the minute in the default theme. Currently looks like this:

![image](https://github.com/MyreMylar/pygame_gui/assets/13382426/e36a47ad-aaeb-412c-967b-274c0b4831e5)

So making it prettier and those functionality unit tests so it can pass the coverage CI are likely the next steps. It's not going to make it into version 0.6.10 - because that is coming out relatively soon-ish, but perhaps in the next version.

GimLala commented:
The tests are failing currently, because of a singular line which updates the dimensions of the `UIAutoResizingContainer `within the `UIScrollingContainer `when the absolute min and max rects are calculated.
The tests worked before because they relied on the scroll_bars appearing when the scrollable area is increased beyond the size of the view container, but because of that line, it automatically detects that it should not be the increased size and resizes back down to the size of the `min_edges_rect`, which is just the original size passed in.
This means that the `set_scrollable_area_dimensions `is useless when `should_grow_automatically` is set to True, which may cause some bugs with previously working applications?

1. Should we make it so that when the `set_scrollable_area_dimensions `is used, we should turn `should_grow_automatically `to False? and/or
2. Should we make it so that when `set_dimensions `is called on a `UIAutoResizingContainer`, we also set `should_update_dimensions` to True?

GimLala commented:
As you can see from the commit above, I started adding the tests to UIForm. You can see a list of todo comments which lists what tests are left to add/complete. Let me know if any tests are needed apart from that

GimLala commented:
The code cov doesn't seem to be running

GimLala commented:
Some tests seem to be failing when we put a UIForm inside a UIForm, but I suspect it has something to do with UIScrollingContainers instead. I'll remove the UIForms within UIForms, but this should be something to look into.

MyreMylar commented:
> The tests are failing currently, because of a singular line which updates the dimensions of the UIAutoResizingContainer within the UIScrollingContainer when the absolute min and max rects are calculated.
> The tests worked before because they relied on the scroll_bars appearing when the scrollable area is increased beyond the size of the view container, but because of that line, it automatically detects that it should not be the increased size and resizes back down to the size of the min_edges_rect, which is just the original size passed in.
> This means that the set_scrollable_area_dimensions is useless when should_grow_automatically is set to True, which may cause some bugs with previously working applications?
> 
> Should we make it so that when the set_scrollable_area_dimensions is used, we should turn should_grow_automatically to False? and/or
> Should we make it so that when set_dimensions is called on a UIAutoResizingContainer, we also set should_update_dimensions to True?

Might be late to the party on replying to this. I guess the spirit of the original auto scrolling container was that it would grow & shrink based on the contents of the auto resizing container - meanwhile the `set_scrolling_dimensions` function sort of ignores that and just sets a size for the scrolling contents.

I think using set_scrolling_dimensions should probably disable the auto growing and shrinking aspect of the auto resizing container - until it is explicitly turned back on (which should also be an option). But we should also update the documentation of the function to say that. 

> Also, should we raise a warnings for certain methods which UIForm inherits from UIScrollingContainer, but aren't intended to be used with UIForm? for example, set_scrollable_area_dimensions

Various ways to handle this:

a) make UIForm have a "has a" relationship to the UIScrolling container rather than an "is a" relationship.
b) override "set_scrollable_area_dimensions" for the UIForm so it doesn't call the base version in UIScrollingContainer and does nothing except issue a warning.
c) make it work OK with the UIForm even if it is a bit dumb?

> Edit: Now it does after the new commit, code cov seems to be slightly buggy

Yeah, it sometimes seems to skip an update to a pull request, I've not figured out exactly why, but it is better than nothing!

GimLala commented:
> I think using set_scrolling_dimensions should probably disable the auto growing and shrinking aspect of the auto resizing container - until it is explicitly turned back on (which should also be an option). But we should also update the documentation of the function to say that.

I agree with that. We can just add a getter and setter for the auto growing option. Alternatively, we could also add the option to pass in -1, -1 in set_scrollable_dimensions to turn it on.

> a) make UIForm have a "has a" relationship to the UIScrolling container rather than an "is a" relationship.

I think I agree with this the least because it doesn't feel write to have to use 2 UIElements to do the work of 1. Also, it makes sense to me that UIForm _is_ a scrolling container, rather than _having_ one. We'd also have to override some functions to call the methods of the contained scrolling container anyway, and that is just extra work for no reward. 
It feels similar to saying whether each sub-class of UIElement could just instead contain a UIElement instance instead of being one itself and just have all of the methods operate on that UIElement instance instead. Sure, it would work, but it makes things more confusing and unintuitive and would definitely lead to unexpected bugs in the future.

Infact, it did raise confusion for me when UISelectionList was just inheriting from UIElement instead of UIContainer or some other class because it clearly does contain other elements. Also, it has a list_and_scroll_bar_container even though I think it should _be_ the list_and_scroll_bar_container. Might be something to look at in the future, perhaps.

> b) override "set_scrollable_area_dimensions" for the UIForm so it doesn't call the base version in UIScrollingContainer and does nothing except issue a warning.

I feel like if someone was to call the function, it's their responsibility for what the function does. I think that would be the expected behaviour as well, instead of just blocking them from the usage of that method. The only such case where we should interfere is when calling the function breaks something about the element or crashing the program. If it is something untested, or something that we don't expect the user to use, or perhaps use a different method, then we should raise a warning, but still call the base method. If there is just one semi-reasonable way of doing something, even if we don't know why that thing would be needed, we don't need to raise a warning or block the way. We just need to try and make sure that the method does what its name suggests it should do.

GimLala commented:
> Infact, it did raise confusion for me when UISelectionList was just inheriting from UIElement instead of UIContainer or some other class because it clearly does contain other elements. Also, it has a list_and_scroll_bar_container even though I think it should be the list_and_scroll_bar_container. Might be something to look at in the future, perhaps.

Also, UIScrollingContainer should _be_ the _root_container. I don't see why we need a root container, because we anyways have to relay the set_position, set_relative_position and set_dimension calls to the root container anyway.
The fact that UIScrollingContainer doesn't even inherit from UIContainer or IUIContainerLikeInterface just feels wrong. Also, since both the UIScrollingContainer and the _root_container are within the parent container, it makes counting the number of elements more confusing.

GimLala commented:
I was checking the recursion errors and one of them happens when the size is set to 50 or lesser in the horizontal axis or 5 or lesser in the vertical axis. I don't understand why though, and there are more errors to find which happen when UIForms are placed within UIForms.

GimLala commented:
I think I will not have much time to work on this to complete the remaining tests unfortunately. I hope I have made sufficient enough progress that you can finish the remaining tests as I have highlighted in some TODO comments at the top of the test class.
however, I may be able to assist if there are weird errors raised by this element.
P.S. Have you seen my comments above?

MyreMylar commented:
I think we will try merging this and see how people get on with it. 

I don't personally have any use cases for it in my existing games or demos so I'm just adding the test program @GimLala posted to the examples repository,
